### PR TITLE
attention refactor: the trilogy 

### PR DESCRIPTION
**Pull request by williamberman**
Replaces `AttentionBlock` with `Attention`. We can just rename the projection layers in the attention block before the state dict is loaded. So this ends up being a straight swap and we don't have to do a deprecation cycle.

See the rest of the PR comments

### CODE REVIEW

File: tests/models/test_layers_utils.py [hunk 1]
```diff
@@ -314,59 +314,6 @@ def test_restnet_with_kernel_sde_vp(self):
         assert torch.allclose(output_slice.flatten(), expected_slice, atol=1e-3)
 
 
-class AttentionBlockTests(unittest.TestCase):
```
Comment (williamberman) [inline tests/models/test_layers_utils.py:317]: I re-configured these tests with the equivalent configuration on `Attention` and manually ran them on test target devices to confirm they still pass. We can just remove the tests

File: src/diffusers/models/modeling_utils.py [hunk 2]
```diff
@@ -761,6 +763,42 @@ def _find_mismatched_keys(
 
         return model, missing_keys, unexpected_keys, mismatched_keys, error_msgs
 
+    @classmethod
+    def convert_deprecated_attention_blocks(cls, state_dict):
```
Comment (williamberman) [inline src/diffusers/models/modeling_utils.py:767]: After we load the state dict, we can just manually rename the layers ({query, key, value, out projection) to the new layer names.

File: src/diffusers/models/attention_processor.py [hunk 3]
```diff
@@ -65,13 +65,18 @@ def __init__(
         out_bias: bool = True,
         scale_qk: bool = True,
         only_cross_attention: bool = False,
+        eps: float = 1e-5,
+        rescale_output_factor: float = 1.0,
+        residual_connection: bool = False,
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:70]: Configurable group norm epsilon, output rescale, and residual connection

File: src/diffusers/models/attention_processor.py [hunk 4]
```diff
@@ -407,10 +412,22 @@ def __call__(
         encoder_hidden_states=None,
         attention_mask=None,
     ):
+        residual = hidden_states
+
+        input_ndim = hidden_states.ndim
+
+        if input_ndim == 4:
+            batch_size, channel, height, width = hidden_states.shape
+            hidden_states = hidden_states.view(batch_size, channel, height * width).transpose(1, 2)
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:421]: allow 2d sequences, by flattening on input and reshaping on output

File: src/diffusers/models/attention_processor.py [hunk 5]
```diff
@@ -407,10 +412,22 @@ def __call__(
...
+
         batch_size, sequence_length, _ = (
             hidden_states.shape if encoder_hidden_states is None else encoder_hidden_states.shape
         )
         attention_mask = attn.prepare_attention_mask(attention_mask, sequence_length, batch_size)
+
+        if attn.group_norm is not None:
+            hidden_states = attn.group_norm(hidden_states.transpose(1, 2)).transpose(1, 2)
+
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:430]: optional group norm for non-added kv attention processors

File: src/diffusers/models/attention_processor.py [hunk 6]
```diff
@@ -434,6 +451,14 @@ def __call__(
         # dropout
         hidden_states = attn.to_out[1](hidden_states)
 
+        if input_ndim == 4:
+            hidden_states = hidden_states.transpose(-1, -2).reshape(batch_size, channel, height, width)
+
+        if attn.residual_connection:
+            hidden_states = hidden_states + residual
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:458]: optional residual connection

File: src/diffusers/models/attention_processor.py [hunk 7]
```diff
@@ -434,6 +451,14 @@ def __call__(
...
+
+        hidden_states = hidden_states / attn.rescale_output_factor
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:460]: add optional output rescaling

**Comment by williamberman** [comment]
Ran integration tests here: https://github.com/huggingface/diffusers/pull/3389 ~same number failing (some can't run due to not having token when on branch)

File: tests/models/test_layers_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply tests/models/test_layers_utils.py:317]: Cool!

File: src/diffusers/models/unet_2d_blocks.py [hunk 8]
```diff
@@ -427,12 +427,16 @@ def __init__(
...
                         in_channels,
-                        num_head_channels=attn_num_head_channels,
+                        heads=in_channels // attn_num_head_channels if attn_num_head_channels is not None else 1,
+                        dim_head=attn_num_head_channels if attn_num_head_channels is not None else in_channels,
                         rescale_output_factor=output_scale_factor,
                         eps=resnet_eps,
                         norm_num_groups=resnet_groups,
+                        residual_connection=True,
+                        bias=True,
+                        upcast_softmax=True,
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:439]:
```suggestion
                        upcast_softmax=True,
                        _deprecated_attention_block=True,
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 9]
```diff
@@ -711,12 +715,16 @@ def __init__(
...
                     out_channels,
-                    num_head_channels=attn_num_head_channels,
+                    heads=out_channels // attn_num_head_channels if attn_num_head_channels is not None else 1,
+                    dim_head=attn_num_head_channels if attn_num_head_channels is not None else out_channels,
                     rescale_output_factor=output_scale_factor,
                     eps=resnet_eps,
                     norm_num_groups=resnet_groups,
+                    residual_connection=True,
+                    bias=True,
+                    upcast_softmax=True,
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:727]:
```suggestion
                    upcast_softmax=True,
                    _deprecated_attention_block=True,
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 10]
```diff
@@ -1060,12 +1068,16 @@ def __init__(
...
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:1080]:
```suggestion
                    upcast_softmax=True,
                    _deprecated_attention_block=True,
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 11]
```diff
@@ -1134,11 +1146,16 @@ def __init__(
...
                     out_channels,
-                    num_head_channels=attn_num_head_channels,
+                    heads=out_channels // attn_num_head_channels if attn_num_head_channels is not None else 1,
+                    dim_head=attn_num_head_channels if attn_num_head_channels is not None else out_channels,
                     rescale_output_factor=output_scale_factor,
                     eps=resnet_eps,
+                    norm_num_groups=32,
+                    residual_connection=True,
+                    bias=True,
+                    upcast_softmax=True,
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:1158]:
```suggestion
                    upcast_softmax=True,
                    _deprecated_attention_block=True,
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 12]
```diff
@@ -1703,12 +1720,16 @@ def __init__(
...
                     out_channels,
-                    num_head_channels=attn_num_head_channels,
+                    heads=out_channels // attn_num_head_channels if attn_num_head_channels is not None else 1,
+                    dim_head=attn_num_head_channels if attn_num_head_channels is not None else out_channels,
                     rescale_output_factor=output_scale_factor,
                     eps=resnet_eps,
                     norm_num_groups=resnet_groups,
+                    residual_connection=True,
+                    bias=True,
+                    upcast_softmax=True,
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:1732]:
```suggestion
                    upcast_softmax=True,
                    _deprecated_attention_block=True,
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 13]
```diff
@@ -2037,12 +2058,16 @@ def __init__(
...
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2070]:
```suggestion
                    upcast_softmax=True,
                    _deprecated_attention_block=True,
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 14]
```diff
@@ -2109,11 +2134,16 @@ def __init__(
...
                 out_channels,
-                num_head_channels=attn_num_head_channels,
+                heads=out_channels // attn_num_head_channels if attn_num_head_channels is not None else 1,
+                dim_head=attn_num_head_channels if attn_num_head_channels is not None else out_channels,
                 rescale_output_factor=output_scale_factor,
                 eps=resnet_eps,
+                norm_num_groups=32,
+                residual_connection=True,
+                bias=True,
+                upcast_softmax=True,
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2146]:
```suggestion
                upcast_softmax=True,
                _deprecated_attention_block=True,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py [hunk 15]
```diff
@@ -709,12 +709,14 @@ def __call__(
...
-        use_xformers = self.vae.decoder.mid_block.attentions[0]._use_memory_efficient_attention_xformers
+        use_torch_2_0_or_xformers = self.vae.decoder.mid_block.attentions[0].processor in [
+            AttnProcessor2_0,
+            XFormersAttnProcessor,
+            LoRAXFormersAttnProcessor,
+        ]
         # if xformers or torch_2_0 is used attention block does not need
         # to be in float32 which can save lots of memory
-        if not use_torch_2_0_attn and not use_xformers:
+        if not use_torch_2_0_or_xformers:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:719]: nice!

File: src/diffusers/models/attention_processor.py (hunk 3 above)
Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:70]:
```suggestion
        residual_connection: bool = False,
        _deprecated_attention_block: bool = False
```

Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:70]:
```suggestion
        residual_connection: bool = False,
        _from_deprecated_attn_block: bool = False
```

File: src/diffusers/models/unet_2d_blocks.py (hunk 13 above)
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2070]:
```suggestion
                    upcast_softmax=True,
                    _from_deprecated_attn_block=True,
```

File: src/diffusers/models/modeling_utils.py [hunk 16]
```diff
@@ -583,6 +583,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                 if device_map is None:
                     param_device = "cpu"
                     state_dict = load_state_dict(model_file, variant=variant)
+                    cls.convert_deprecated_attention_blocks(state_dict)
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:586]:
```suggestion
                    cls.convert_deprecated_attention_blocks(model, state_dict)
```

File: src/diffusers/models/modeling_utils.py [hunk 17]
```diff
@@ -625,6 +626,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                 model = cls.from_config(config, **unused_kwargs)
 
                 state_dict = load_state_dict(model_file, variant=variant)
+                cls.convert_deprecated_attention_blocks(state_dict)
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:629]:
```suggestion
                cls.convert_deprecated_attention_blocks(model, state_dict)
```

File: src/diffusers/models/modeling_utils.py (hunk 2 above)
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:767]:
```suggestion
    def convert_deprecated_attention_blocks(cls, model, state_dict):
```

File: src/diffusers/models/modeling_utils.py [hunk 18]
```diff
@@ -761,6 +763,42 @@ def _find_mismatched_keys(
...
+        # We check for the deprecated attention block via the `proj_attn` layer in the state dict
+        # The only other class with a layer called `proj_attn` is in `SelfAttention1d` which is used
+        # by only the top level model, `UNet1DModel`. Since, `UNet1DModel` wont have any of the deprecated
+        # attention blocks, we can just early return.
+        if cls.__name__ == "UNet1DModel":
+            return
+
+        deprecated_attention_block_paths = []
+
+        for k in state_dict.keys():
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:777]: This check is a bit brittle IMO. As shown in the suggestions above, we can attach a `_from_deprecated_attn_block` attribute to all attn classes that come from the deprecated block. Then we Breath First Search over the modules and whenever we hit a `_from_deprecated_attn_block`

```suggestion
        def recursive_find_attn_block(name, module, deprecated):
            if hasattr(module, "_from_deprecated_attn_block") and module._from_deprecated_attn_block:
                deprecated.append(name)
            for sub_name, sub_module in module.named_children():
                sub_name = sub_name if name == "" else f"{name}.{sub_name}" 
                recursive_find_attn_block(sub_name, sub_module, deprecated)
                
        recursive_find_attn_block("", model, deprecated_attention_block_paths)
```

File: src/diffusers/models/modeling_utils.py [hunk 19]
```diff
@@ -761,6 +763,42 @@ def _find_mismatched_keys(
...
         return model, missing_keys, unexpected_keys, mismatched_keys, error_msgs
 
+    @classmethod
+    def convert_deprecated_attention_blocks(cls, state_dict):
+        # We check for the deprecated attention block via the `proj_attn` layer in the state dict
+        # The only other class with a layer called `proj_attn` is in `SelfAttention1d` which is used
+        # by only the top level model, `UNet1DModel`. Since, `UNet1DModel` wont have any of the deprecated
+        # attention blocks, we can just early return.
+        if cls.__name__ == "UNet1DModel":
+            return
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:773]:
```suggestion
```

That's a bit too brittle - what if a class is not called `UNet1DModel` but still uses the UNet1DModel block?

File: src/diffusers/models/modeling_utils.py [hunk 20]
```diff
@@ -761,6 +763,42 @@ def _find_mismatched_keys(
...
+
+        deprecated_attention_block_paths = []
+
+        for k in state_dict.keys():
+            if "proj_attn.weight" in k:
+                index = k.index("proj_attn.weight")
+                path = k[: index - 1]
+                deprecated_attention_block_paths.append(path)
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:781]:
```suggestion
```

Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:781]:
```suggestion
```

**Review by patrickvonplaten — COMMENTED** [review]
Super this looks very good to me! Much easier now the PR. I added some suggestions on how we can be a bit safer that the conversion script really only hits classes that need conversion

File: src/diffusers/models/modeling_utils.py (hunk 18 above)
Comment (williamberman) [reply src/diffusers/models/modeling_utils.py:777]: parfait!

**Review by patrickvonplaten — APPROVED** [review]
Good to go for me!

File: src/diffusers/models/attention_processor.py (hunk 3 above)
Comment (sayakpaul) [reply src/diffusers/models/attention_processor.py:70]: At some point (if need be), we should maybe also serialize them in the config.json of UNet. WDYT?

File: src/diffusers/models/attention_processor.py (hunk 7 above)
Comment (sayakpaul) [reply src/diffusers/models/attention_processor.py:465]: I don't see the conditional branch :/

File: src/diffusers/models/attention_processor.py [hunk 21]
```diff
@@ -858,11 +947,22 @@ def __init__(self, hidden_size, cross_attention_dim, rank=4, attention_op: Optio
         self.to_out_lora = LoRALinearLayer(hidden_size, hidden_size, rank)
 
     def __call__(self, attn: Attention, hidden_states, encoder_hidden_states=None, attention_mask=None, scale=1.0):
+        residual = hidden_states
+
+        input_ndim = hidden_states.ndim
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:952]: Can we remove these whitespaces here and elsewhere?

**Review by sayakpaul — APPROVED** [review]
Smooth! 

I would like to give this another review on Monday or Tuesday. But that's non-blocking. Totally fine for me to merge.

File: src/diffusers/models/attention_processor.py (hunk 7 above)
Comment (williamberman) [reply src/diffusers/models/attention_processor.py:465]: rescale output factor defaults to 1

File: src/diffusers/models/attention_processor.py (hunk 3 above)
Comment (williamberman) [reply src/diffusers/models/attention_processor.py:70]: if needed :)

**Pull request merged**